### PR TITLE
Adds timeout capabilities to blocking Rpc Calls

### PR DESCRIPTION
**Pull request by ThadHouse**

### CODE REVIEW

File: include/ntcore_c.h [hunk 1]
```diff
@@ -305,6 +305,8 @@ int NT_NotifierDestroyed();
 /*
  * Remote Procedure Call Functions
  */
+ 
+#define TIMEOUT_INDEFINITE -1
```
Comment (PeterJohnson) [inline include/ntcore_c.h]: This needs an NT prefix.  Is this really necessary given the non-timeout variants?

Comment (ThadHouse) [inline include/ntcore_c.h]: Yeah its probably not actually needed. I'll remove it.

File: src/Storage.cpp [hunk 2]
```diff
@@ -1407,15 +1407,32 @@ unsigned int Storage::CallRpc(StringRef name, StringRef params) {
...
     auto i =
         m_rpc_results.find(std::make_pair(call_uid >> 16, call_uid & 0xffff));
     if (i == m_rpc_results.end()) {
       if (!blocking || m_terminating) return false;
-      m_rpc_results_cond.wait(lock);
+      if (time_out == kTimeout_Indefinite) {
+        m_rpc_results_cond.wait(lock);
+      } else {
+        auto time_out_millis = 
+          std::chrono::milliseconds(static_cast<long long>(time_out * 1000));
```
Comment (PeterJohnson) [inline src/Storage.cpp]: Why do this instead of just converting directly to seconds?

File: src/RpcServer.h [hunk 3]
```diff
@@ -9,6 +9,7 @@
 #define NT_RPCSERVER_H_
 
 #include <atomic>
+#include <chrono>
```
Comment (PeterJohnson) [inline src/RpcServer.h]: Shouldn't need chrono here.. should be in the C++ file instead?

File: src/Storage.h [hunk 4]
```diff
@@ -9,6 +9,7 @@
 #define NT_STORAGE_H_
 
 #include <atomic>
+#include <chrono>
```
Comment (PeterJohnson) [inline src/Storage.h]: Shouldn't need chrono here.

File: src/ntcore_c.cpp [hunk 5]
```diff
@@ -8,6 +8,7 @@
 #include "ntcore.h"
 
 #include <cassert>
+#include <chrono>
```
Comment (PeterJohnson) [inline src/ntcore_c.cpp]: Shouldn't need chrono here.

File: src/RpcServer.cpp [hunk 6]
```diff
@@ -61,10 +61,26 @@ void RpcServer::ProcessRpc(StringRef name, std::shared_ptr<Message> msg,
...
 bool RpcServer::PollRpc(bool blocking, RpcCallInfo* call_info) {
+  return PollRpc(blocking, kTimeout_Indefinite, call_info);
+}
+
+bool RpcServer::PollRpc(bool blocking, double time_out, RpcCallInfo* call_info) {
   std::unique_lock<std::mutex> lock(m_mutex);
   while (m_poll_queue.empty()) {
     if (!blocking || m_terminating) return false;
-    m_poll_cond.wait(lock);
+    if (time_out == kTimeout_Indefinite) {
```
Comment (PeterJohnson) [inline src/RpcServer.cpp]: Should this be < 0 instead of == kTimeout_Indefinite?

File: include/ntcore_cpp.h [hunk 7]
```diff
@@ -9,6 +9,7 @@
 #define NTCORE_CPP_H_
 
 #include <cassert>
+#include <chrono>
```
Comment (PeterJohnson) [inline include/ntcore_cpp.h]: Don't need chrono here.

File: src/RpcServer.h (hunk 3 above)
Comment (ThadHouse) [inline src/RpcServer.h]: These chronos where an initial version where the functions took a std::milliseconds value. I'll fix them.

File: ntcore.def [hunk 8]
```diff
@@ -90,3 +90,5 @@ NT_SetDefaultEntryRaw @91
 NT_SetDefaultEntryBooleanArray @92
 NT_SetDefaultEntryDoubleArray @93
 NT_SetDefaultEntryStringArray @94
+NT_PollRpcTimeout @95
+NT_GetRpcResultTimeout @96
```
Comment (PeterJohnson) [inline ntcore.def]: No EOL at EOF..

**Comment by ThadHouse** [comment]
Its added. We should make the formatting tools work over here at some point.

**Pull request merged**